Modix.Services/Promotions/PromotionsService.cs

```diff
@@ -245,6 +255,52 @@ public async Task AddCommentAsync(long campaignId, PromotionSentiment sentiment,
             PublishActionNotificationAsync(resultAction);
         }
 
+        public async Task<bool> VotePromoAsync(long campaignId,IGuildUser user, PromotionSentiment sentiment)
```

Comment:
This function was written because I dont like exceptions being thrown for vote failures here. I would consider that expected and therefore not exceptional. we would probably need to return a string or an error object here to allow for updating the embed

Reply:
I see no need to duplicate logic here. We already have `AddCommentAsync()` and `UpdateCommentAsync()`. We should either have this method simply call those two, depending on whether a comment already exists or not (in which case, this should be `AddOrUpdateCommendAsync()`), OR refactor those two entirely into something like `SetCommentAsync()`.

And you're gonna have to clarify what you mean about exceptions. You just mean that when there's a permissions failure or a rank failure or whatever, we shouldn't throw an exception. Generally, yes, I'd agree, but that is the current mechanism by which we get error messages propagated up to the client. It is not ideal, but that's the way it is now, and changing that I'd say is outside of the scope of this feature. With your implementation here, you're actually NOT propagating error information up to the frontend. So, when a failure occurs, you have no way of notifying the user about WHY. Which is important in a UI.

Reply:
The error propagation is why i suggested an error object to be returned instead, with the exception approach it means you have to wrap the call in a try catch in PromotionDialogueBehavior and then display the exception text to the user, because we don't want Modix to log every failed vote attempt as an exception in modix_log OR send a message in the poll channel  as that will clutter it up quickly. From my testing allowing the exceptions to bubble up organically was NOTABLY slower then returning a string to the behavior. However the difference in time between wrapping the call in a try  and adding the exception text to the embed, and just returning a string was miniscule. So the try seems to be the way to go if we want to keep the call with the exceptions. Though i will still advocate for an error object return as an option, as the exception approach seems needlessly heavy and like we are using exceptions for unexceptional things.

Reply:
>  it means you have to wrap the call in a try catch in PromotionDialogueBehavior and then display the exception text to the user

You should really be doing this regardless, considering your behavior serves as an "entry point" for user interaction. This is part of why I would like to see a "generic" dialog implementation in the future.

> From my testing allowing the exceptions to bubble up organically was NOTABLY slower then returning a string to the behavior.

No way is `throw`ing a single exception actually the cause of noticable lag.

> Though i will still advocate for an error object return as an option, as the exception approach seems needlessly heavy and like we are using exceptions for unexceptional things

Yes, you are 100% correct. All the maintainers agree. However, I don't think it's worth re-implementing logic that already exists, just for the sake of avoiding exception-based flow, and I also think it's way out of scope to try and refactor the existing methods to not throw, cause then you have to refactor all the other features that use those methods.

Adopting non-exception-based flow control across the board should be a separate refactoring effort.

Reply:
> No way is throwing a single exception actually the cause of noticable lag.
Noticeable lag between the reaction being made and the failure reason bubbling up to the top. When the call is wrapped there is very little difference. 

> Adopting non-exception-based flow control across the board should be a separate refactoring effort.
fair enough, ill switch to the already in use method